File: teedinamo/static/js/pos.js
let cart = [];
let lastTransNo = "";
let searchTimer = null;

const searchInput  = document.getElementById("searchInput");
const searchResults= document.getElementById("searchResults");
const cartEl       = document.getElementById("cart");
const paymentInput = document.getElementById("paymentInput");

// ─── Search ───────────────────────────────────────────────
searchInput.addEventListener("keydown", e => {
  if (e.key === "Enter") {
    const val = searchInput.value.trim();
    if (!val) return;
    // Try barcode first (exact, no spaces)
    if (!val.includes(" ") && val.length >= 4) {
      fetch(`/api/items/barcode/${encodeURIComponent(val)}`)
        .then(r => r.ok ? r.json() : null)
        .then(item => {
          if (item) { addToCart(item); searchInput.value = ""; searchResults.innerHTML = ""; }
          else doSearch(val);
        });
    } else {
      doSearch(val);
    }
  }
});

searchInput.addEventListener("input", () => {
  clearTimeout(searchTimer);
  const q = searchInput.value.trim();
  if (q.length < 2) { searchResults.innerHTML = ""; return; }
  searchTimer = setTimeout(() => doSearch(q), 250);
});

function doSearch(q) {
  fetch(`/api/items/search?q=${encodeURIComponent(q)}`)
    .then(r => r.json())
    .then(renderSearchResults);
}

function renderSearchResults(items) {
  if (!items.length) {
    searchResults.innerHTML = `<div style="padding:12px;color:#999">ไม่พบสินค้า</div>`;
    return;
  }
  searchResults.innerHTML = items.map(it => `
    <div class="search-item" onclick="addToCart(${JSON.stringify(it).replace(/"/g,'&quot;')})">
      <div>
        <div class="item-name">${it.itemname}</div>
        <div class="item-meta">${it.itemid} | ${it.barcode || '-'} | ${it.unit}
          ${it.stock <= 5 ? `<span class="item-stock-low">⚠ สต็อก ${it.stock}</span>` : `| สต็อก ${it.stock}`}
        </div>
      </div>
      <div class="item-price">฿${Number(it.price).toLocaleString('th-TH',{minimumFractionDigits:2})}</div>
    </div>`).join("");
}

// ─── Cart ─────────────────────────────────────────────────
function addToCart(item) {
  const existing = cart.find(c => c.item_id === item.id);
  if (existing) {
    existing.qty++;
  } else {
    cart.push({
      item_id: item.id,
      itemname: item.itemname,
      unit: item.unit,
      unit_price: item.price,
      qty: 1,
      exchange_used: false,
      exchange_value: item.exchange_value || 0,
    });
  }
  renderCart();
  searchInput.focus();
}

function renderCart() {
  if (!cart.length) {
    cartEl.innerHTML = `<div style="color:#999;text-align:center;padding:20px">ตะกร้าว่าง</div>`;
    updateSummary();
    return;
  }
  cartEl.innerHTML = cart.map((it, i) => `
    <div class="cart-item">
      <div>
        <div class="cart-item-name">${it.itemname}</div>
        <div class="cart-item-sub">฿${it.unit_price.toFixed(2)} / ${it.unit}</div>
        ${it.exchange_value > 0 ? `
        <label class="exchange-check">
          <input type="checkbox" ${it.exchange_used ? 'checked' : ''}
            onchange="toggleExchange(${i}, this.checked)">
          มีแบตเตอรี่เก่า (หัก ฿${it.exchange_value.toFixed(2)})
        </label>` : ''}
      </div>
      <div class="cart-qty">
        <button onclick="changeQty(${i},-1)">−</button>
        <input type="number" value="${it.qty}" min="1"
          onchange="setQty(${i}, this.value)" style="width:44px">
        <button onclick="changeQty(${i},1)">+</button>
      </div>
      <div>
        <div class="cart-item-price">฿${lineTotal(it).toFixed(2)}</div>
        <div class="cart-item-remove" onclick="removeItem(${i})">✕</div>
      </div>
    </div>`).join("");
  updateSummary();
}

function lineTotal(it) {
  return it.unit_price * it.qty - (it.exchange_used ? it.exchange_value : 0);
}

function changeQty(i, d) {
  cart[i].qty = Math.max(1, cart[i].qty + d);
  renderCart();
}
function setQty(i, v) {
  cart[i].qty = Math.max(1, parseInt(v) || 1);
  renderCart();
}
function removeItem(i) {
  cart.splice(i, 1);
  renderCart();
}
function toggleExchange(i, v) {
  cart[i].exchange_used = v;
  renderCart();
}
function clearCart() {
  cart = [];
  renderCart();
}

function updateSummary() {
  const subtotal  = cart.reduce((s, it) => s + it.unit_price * it.qty, 0);
  const deduct    = cart.reduce((s, it) => s + (it.exchange_used ? it.exchange_value : 0), 0);
  const grand     = subtotal - deduct;
  const payment   = parseFloat(paymentInput.value) || 0;
  const change    = payment - grand;

  document.getElementById("subtotal").textContent     = `฿${subtotal.toFixed(2)}`;
  document.getElementById("exchangeDeduct").textContent = `-฿${deduct.toFixed(2)}`;
  document.getElementById("grandTotal").textContent   = `฿${grand.toFixed(2)}`;
  document.getElementById("changeAmt").textContent    = change >= 0 ? `฿${change.toFixed(2)}` : "-";
}

paymentInput.addEventListener("input", updateSummary);

function setCash(amount) {
  paymentInput.value = amount;
  updateSummary();
}

// ─── Checkout ─────────────────────────────────────────────
function checkout() {
  if (!cart.length) { alert("ไม่มีสินค้าในตะกร้า"); return; }
  const payment = parseFloat(paymentInput.value) || 0;
  const grand = cart.reduce((s, it) => s + lineTotal(it), 0);
  if (payment < grand) { alert("กรุณาใส่จำนวนเงินให้ครบ"); paymentInput.focus(); return; }

  fetch("/api/checkout", {
    method: "POST",
    headers: {"Content-Type": "application/json"},
    body: JSON.stringify({ items: cart, payment }),
  })
  .then(r => r.json())
  .then(data => {
    if (data.error) { alert(data.error); return; }
    lastTransNo = data.trans_no;
    document.getElementById("modalTransNo").textContent = `บิลเลขที่: ${data.trans_no}`;
    document.getElementById("modalTotal").textContent   = `ยอดรวม: ฿${data.total.toFixed(2)}`;
    document.getElementById("modalChange").textContent  = `ทอน: ฿${data.change.toFixed(2)}`;
    document.getElementById("receiptModal").classList.remove("hidden");
    clearCart();
    paymentInput.value = "";
  })
  .catch(() => alert("เกิดข้อผิดพลาด"));
}

function printReceipt() {
  window.open(`/receipt/${lastTransNo}`, "_blank");
}

function closeModal() {
  document.getElementById("receiptModal").classList.add("hidden");
  searchInput.focus();
}

// ─── Keyboard shortcuts ───────────────────────────────────
document.addEventListener("keydown", e => {
  if (e.key === "F2") { searchInput.focus(); e.preventDefault(); }
  if (e.key === "F12") { checkout(); e.preventDefault(); }
  if (e.key === "Escape") { closeModal(); }
});

// init
renderCart();

File: teedinamo/routes/pos.py
from flask import Blueprint, render_template, request, jsonify, session
from routes.auth import login_required
from models.item import search_items, get_item_by_barcode, get_item_by_id, get_price
from models.transaction import create_transaction, get_transaction

pos_bp = Blueprint("pos", __name__)


@pos_bp.route("/pos")
@login_required
def pos():
    return render_template("pos.html",
                           username=session["username"],
                           price_level=session["price_level"])


@pos_bp.route("/api/items/search")
@login_required
def api_search():
    q = request.args.get("q", "").strip()
    if not q:
        return jsonify([])
    items = search_items(q)
    price_level = session["price_level"]
    result = []
    for it in items:
        result.append({
            "id": it["id"],
            "itemid": it["itemid"],
            "itemname": it["itemname"],
            "barcode": it["barcode"],
            "unit": it["unit"],
            "stock": it["stock"],
            "price": get_price(it, price_level),
            "exchange_value": it["exchange_value"],
        })
    return jsonify(result)


@pos_bp.route("/api/items/barcode/<barcode>")
@login_required
def api_barcode(barcode):
    item = get_item_by_barcode(barcode)
    if not item:
        return jsonify({"error": "ไม่พบสินค้า"}), 404
    price_level = session["price_level"]
    return jsonify({
        "id": item["id"],
        "itemid": item["itemid"],
        "itemname": item["itemname"],
        "barcode": item["barcode"],
        "unit": item["unit"],
        "stock": item["stock"],
        "price": get_price(item, price_level),
        "exchange_value": item["exchange_value"],
    })


@pos_bp.route("/api/checkout", methods=["POST"])
@login_required
def api_checkout():
    data = request.json
    items = data.get("items", [])
    payment = float(data.get("payment", 0))
    if not items:
        return jsonify({"error": "ไม่มีสินค้าในตะกร้า"}), 400

    cart = []
    for it in items:
        item = get_item_by_id(it["item_id"])
        if not item:
            return jsonify({"error": f"ไม่พบสินค้า id={it['item_id']}"}), 400
        cart.append({
            "item_id": it["item_id"],
            "qty": float(it["qty"]),
            "price_level": session["price_level"],
            "unit_price": float(it["unit_price"]),
            "exchange_used": bool(it.get("exchange_used", False)),
            "exchange_value": float(it.get("exchange_value", 0)),
        })

    total = sum(
        (i["unit_price"] * i["qty"]) - (i["exchange_value"] if i["exchange_used"] else 0)
        for i in cart
    )
    if payment < total:
        return jsonify({"error": "รับเงินน้อยกว่ายอดรวม"}), 400

    trans_no, total, change = create_transaction(session["user_id"], cart, payment)
    return jsonify({"trans_no": trans_no, "total": total, "change": change})


@pos_bp.route("/receipt/<trans_no>")
@login_required
def receipt(trans_no):
    t, items = get_transaction(trans_no)
    if not t:
        return "ไม่พบใบเสร็จ", 404
    return render_template("receipt.html", t=t, items=items)
